### ISSUE TITLE
Getting "Unparseable date" when using filters

### ISSUE THREAD

**Issue by MattRogish** [ruckus/quickeebooks#26]
Strangely, since upgrading from 0.9 we're now getting "Unparseable date" errors from Quickbooks Online DateTime filters.

I checked, and what is being passed appears to be correct:

`Filter=LastUpdatedTime :AFTER: 2013-01-18T01:29:38+0000`

Except Intuit is responding:

`CUSTOMER QUERY query is invalid. Message: Unparseable date: "2013-01-18T01:29:38 0000"`

I find the error goes away if I switch to the time zone "string" format (as specified here) https://ipp.developer.intuit.com/0010_Intuit_Partner_Platform/0050_Data_Services/0400_QuickBooks_Online/0100_Calling_Data_Services/0030_Retrieving_Objects

by changing the filter date format to use a capital Z:

```
    `DATE_TIME_FORMAT = '%Y-%m-%dT%H:%M:%S%Z'`
```

(instead of  ...%z')

which sends

`Filter=LastUpdatedTime :AFTER: 2013-01-18T01:31:24UTC`

Is anyone else able to reproduce my issue? Is this some sort of weird bug on Intuit's side? (the upgrade from 0.9 may have been coincidence)

Thanks,
## 

Matt

**Comment by ruckus** [comment]
Odd, I cannot reproduce this at all. Both of these work for me using `0.1.1`:

``` ruby
s = Quickeebooks::Online::Service::Invoice.new
s.access_token = token
s.realm_id = realm_id
datetime = Time.mktime(2012, 10, 15)
f = Quickeebooks::Online::Service::Filter.new(:datetime, :field => 'CreateTime', :after => datetime)
filters = [f]
s.list(filters)

c = Quickeebooks::Online::Service::Customer.new
c.access_token = token
c.realm_id = realm_id
datetime = Time.mktime(2012, 10, 15)
f = Quickeebooks::Online::Service::Filter.new(:datetime, :field => 'LastUpdatedTime', :after => datetime)
filters = [f]
c.list(filters)
```

In both cases the filter POST param looks like:

```
Filter  LastUpdatedTime :AFTER: 2012-10-15T00:00:00-0700
```

**Comment by MattRogish** [comment]
I forgot to note, we're on 1.9.3, although not sure why that would make a difference. :) 

Also, we're using Rails3 ActiveSupport::TimeWithZone, e.g Time.zone.now

**Comment by ruckus** [comment]
Yeah I am on Ruby 1.9.3 as well, but yes, shouldn't make a difference.

Whats the full code snippet you're working with (like what I have above). Can you please paste it in?

**Comment by ruckus** [comment]
Yeah I can see that with an input of an instance of `ActiveSupport::TimeWithZone` it acts funky.

Apparently `ActiveSupport::TimeWithZone#strftime` does something funky with `%Z` and `%z`:

http://api.rubyonrails.org/classes/ActiveSupport/TimeWithZone.html#method-i-strftime

**Comment by ruckus** [comment]
This patch fixes it:

```
diff --git a/lib/quickeebooks/shared/service/filter.rb b/lib/quickeebooks/shared/service/filter.rb
index ee6d525..f887f32 100644
--- a/lib/quickeebooks/shared/service/filter.rb
+++ b/lib/quickeebooks/shared/service/filter.rb
@@ -5,6 +5,7 @@ module Quickeebooks

         DATE_FORMAT = '%Y-%m-%d'
         DATE_TIME_FORMAT = '%Y-%m-%dT%H:%M:%S%z'
+        ACTIVE_SUPPORT_TIME_WITH_ZONE_FORMAT = '%Y-%m-%dT%H:%M:%S%Z'

         attr_reader :type
         attr_accessor :field, :value
@@ -84,6 +85,8 @@ module Quickeebooks
         def formatted_time(time)
           if time.is_a?(Date)
             time.strftime(DATE_FORMAT)
+          elsif time.is_a?(ActiveSupport::TimeWithZone)
+            time.strftime(ACTIVE_SUPPORT_TIME_WITH_ZONE_FORMAT)
           elsif time.is_a?(DateTime) || time.is_a?(Time)
             time.strftime(DATE_TIME_FORMAT)
           end
```

Give it a shot and let me know?

**Comment by MattRogish** [comment]
Almost. The problem is:

Time.zone.new.is_a? Time => true, so it never reaches the bottom one. 

Could move it higher in the if/elsif chain, or just change the default to be "Big Z". The Quickbooks API states that UTC in words is just fine, although I hesitate to make such a sweeping change without really understanding if it might impact something else. 

## 

Matt

**Comment by ruckus** [comment]
Hmm, I see. Also it was kind of lame having to hard-code the class check of a totally 3rd-party library (ActiveSupport).

This works and is a little more generic. Your idea of using `%Z` I think is OK - if Intuit supports it then so be it.

```
diff --git a/lib/quickeebooks/shared/service/filter.rb b/lib/quickeebooks/shared/service/filter.rb
index ee6d525..b00b341 100644
--- a/lib/quickeebooks/shared/service/filter.rb
+++ b/lib/quickeebooks/shared/service/filter.rb
@@ -4,7 +4,7 @@ module Quickeebooks
       class Filter

         DATE_FORMAT = '%Y-%m-%d'
-        DATE_TIME_FORMAT = '%Y-%m-%dT%H:%M:%S%z'
+        DATE_TIME_FORMAT = '%Y-%m-%dT%H:%M:%S%Z'

         attr_reader :type
         attr_accessor :field, :value
@@ -84,7 +84,7 @@ module Quickeebooks
         def formatted_time(time)
           if time.is_a?(Date)
             time.strftime(DATE_FORMAT)
-          elsif time.is_a?(DateTime) || time.is_a?(Time)
+          elsif time.respond_to?(:strftime) # catch any other Time-like object
             time.strftime(DATE_TIME_FORMAT)
           end
         end
```

**Comment by MattRogish** [comment]
great, yeah, that'll do it I think :)

**Comment by MattRogish** [comment]
:+1:

**Comment by ruckus** [comment]
Released `0.1.2`

**Comment by MattRogish** [comment]
Thanks !!!!!!!
